lib/rucio/core/transfer.py

```diff
@@ -721,381 +1058,239 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
-                                            activity=activity)
-                dest_url = __rewrite_dest_url(dest_url, dest_sign_url=dest_sign_url, dest_scheme=dest_scheme)
-
-                # Get dest space token
-                dest_spacetoken = None
-                if dest_protocol.attributes and 'extended_attributes' in dest_protocol.attributes and \
-                        dest_protocol.attributes['extended_attributes'] and 'space_token' in dest_protocol.attributes['extended_attributes']:
-                    dest_spacetoken = dest_protocol.attributes['extended_attributes']['space_token']
-
-                use_ipv4 = ctx.rse_attrs(source_rse_id).get('use_ipv4', False) or ctx.rse_attrs(dest_rse_id).get('use_ipv4', False)
-
-                # get external_host + strict_copy + archive timeout
-                strict_copy = ctx.rse_attrs(dest_rse_id).get('strict_copy', False)
-                fts_hosts = ctx.rse_attrs(dest_rse_id).get('fts', None)
-                archive_timeout = ctx.rse_attrs(dest_rse_id).get('archive_timeout', None)
-                if source_sign_url == 'gcs':
-                    fts_hosts = ctx.rse_attrs(source_rse_id).get('fts', None)
-                source_globus_endpoint_id = ctx.rse_attrs(source_rse_id).get('globus_endpoint_id', None)
-                dest_globus_endpoint_id = ctx.rse_attrs(dest_rse_id).get('globus_endpoint_id', None)
-
-                if transfertool == 'fts3' and not fts_hosts:
-                    logger(logging.ERROR, 'Destination RSE %s FTS attribute not defined - SKIP REQUEST %s', dest_rse_name, req_id)
-                    continue
-                if transfertool == 'globus' and (not dest_globus_endpoint_id or not source_globus_endpoint_id):
-                    logger(logging.ERROR, 'Destination RSE %s Globus endpoint attributes not defined - SKIP REQUEST %s', dest_rse_name, req_id)
-                    continue
-                if retry_count is None:
-                    retry_count = 0
-                external_host = ''
-                if fts_hosts:
-                    fts_list = fts_hosts.split(",")
-                    external_host = fts_list[0]
-
-                if retry_other_fts:
-                    external_host = fts_list[retry_count % len(fts_list)]
+                allowed_source_rses = [x['id'] for x in parsed_rses]
+
+        filtered_sources = rws.sources
+        # Only keep allowed sources
+        if allowed_source_rses is not None:
+            filtered_sources = filter(lambda s: s.rse_id in allowed_source_rses, filtered_sources)
+        filtered_sources = filter(lambda s: s.rse_name is not None, filtered_sources)
+        # Ignore blacklisted RSEs
+        filtered_sources = filter(lambda s: s.rse_id not in unavailable_read_rse_ids, filtered_sources)
+        # Ignore tape sources if they are not desired
+        filtered_sources = list(filtered_sources)
+        had_tape_sources = len(filtered_sources) > 0
```

Comment:
check the usage of `had_tape_sources` bellow and it will become clear. It's used like "if not filtered_sources and had_tape_sources .... then". Do you have a better idea on how to achieve this goal ?